Author: Andrew Lee (李健秋) <ajqlee@debian.org>
Description: Change default terminal and browser.

diff -Naurp lxpanel.orig/data/default/panels/panel.in lxpanel/data/default/panels/panel.in
--- lxpanel.orig/data/default/panels/panel.in	2011-07-29 14:37:01.000000000 +0200
+++ lxpanel/data/default/panels/panel.in	2012-04-20 14:36:55.535074244 +0200
@@ -53,10 +53,10 @@ Plugin {
             id=pcmanfm.desktop
         }
         Button {
-            id=gnome-terminal.desktop
+            id=lxde-x-terminal-emulator.desktop
         }
         Button {
-            id=firefox.desktop
+            id=lxde-x-www-browser.desktop
         }
     }
 }
